File: problem/ai-battle-series/03-grade-decrypt/src/main.py
import csv
import sys


def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    it = iter(data)
    k = int(next(it))
    if k <= 0:
        print(0)
        return
    subjects = [next(it) for _ in range(k)]
    subject_set = set(subjects)
    need_help = {s: set() for s in subjects}

    try:
        with open("grades.csv", newline="") as f:
            reader = csv.DictReader(f)
            for row in reader:
                subject = row.get("subject", "")
                status = row.get("status", "")
                student = row.get("student_id", "")
                if subject in subject_set and status == "NEED_HELP":
                    need_help[subject].add(student)
    except FileNotFoundError:
        print(0)
        return

    if not subjects:
        print(0)
        return

    result = []
    for student in need_help[subjects[0]]:
        if all(student in need_help[subj] for subj in subjects[1:]):
            result.append(student)

    result.sort()
    print(len(result))
    for student in result:
        print(student)


if __name__ == "__main__":
    main()

File: problem/ai-battle-series/03-grade-decrypt/src/main.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int k = 0;
    if (!(cin >> k)) {
        return 0;
    }
    if (k <= 0) {
        cout << 0 << "\n";
        return 0;
    }

    vector<string> subjects(k);
    unordered_set<string> subject_set;
    subject_set.reserve(k * 2);
    for (int i = 0; i < k; i++) {
        cin >> subjects[i];
        subject_set.insert(subjects[i]);
    }

    unordered_map<string, unordered_set<string>> need_help;
    need_help.reserve(k * 2);
    for (const auto &subj : subjects) {
        need_help.emplace(subj, unordered_set<string>());
    }

    ifstream fin("grades.csv");
    if (!fin) {
        cout << 0 << "\n";
        return 0;
    }

    string line;
    getline(fin, line);
    while (getline(fin, line)) {
        if (line.empty()) {
            continue;
        }
        stringstream ss(line);
        string student;
        string subject;
        string score;
        string status;
        getline(ss, student, ',');
        getline(ss, subject, ',');
        getline(ss, score, ',');
        getline(ss, status, ',');
        if (student.empty() || subject.empty() || status.empty()) {
            continue;
        }
        if (status == "NEED_HELP" && subject_set.count(subject)) {
            need_help[subject].insert(student);
        }
    }

    vector<string> result;
    const auto &first_set = need_help[subjects[0]];
    for (const auto &student : first_set) {
        bool ok = true;
        for (int i = 1; i < k; i++) {
            if (!need_help[subjects[i]].count(student)) {
                ok = false;
                break;
            }
        }
        if (ok) {
            result.push_back(student);
        }
    }

    sort(result.begin(), result.end());
    cout << result.size() << "\n";
    for (const auto &student : result) {
        cout << student << "\n";
    }
    return 0;
}
